src/dataflow/src/source/mod.rs

```diff
@@ -155,15 +157,30 @@ pub(crate) struct SourceData {
     pub(crate) upstream_time_millis: Option<i64>,
 }
 
-#[derive(Debug, Default, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
 /// The output of the decoding operator
+#[derive(Debug, Default, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
 pub struct DecodeResult {
     /// The decoded key
     pub key: Option<Result<Row, DecodeError>>,
     /// The decoded value
     pub value: Option<Result<Row, DecodeError>>,
     /// The index of the decoded value in the stream
     pub position: Option<i64>,
+    /// If this is a Kafka stream, the appropriate metadata
+    // TODO(bwm): This should probably be statically different for different streams, or we should
+    // propagate whether metadata is requested into the decoder
+    pub metadata: Option<KafkaMetadata>,
```

Comment:
Does this need to be a `KafkaMetadata` struct? I saw that for upsert you introduced an `UpsertData` struct that has a `Row` for its metadata, which sounds what we should have here too so that all envelopes operate on the same input of a stream of `DecodeResult`s. Envelopes can look at the RelationDesc to understand what's inside the metadata Row if needed

Reply:
this is gone now.